walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,206 @@
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    long meetingDurationMinutes = request.getDuration();
+
+    // If the meeting duration is over 24 hours, then return no available timeRange.
+    if (meetingDurationMinutes > TimeRange.WHOLE_DAY.duration()) {
+      return Collections.emptyList();
+    }
+
+    // If no Attendees are passed in, return a timeRange containing the whole day.
+    if (request.getAttendees().isEmpty() && request.getOptionalAttendees().isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // Store unavailable TimeRanges for mandatory and optional attendees separately.
+    List<TimeRange> unavailableMandatoryTimeRanges =
+        getUnavailableTimeRanges(events, mandatoryAttendees);
+    List<TimeRange> unavailableOptionalTimeRanges =
+        getUnavailableTimeRanges(events, optionalAttendees);
+
+    // Get available TimeRanges from unavailable TimeRanges.
+    List<TimeRange> availableMandatoryTimeRanges =
+        getAvailableTimeRanges(unavailableMandatoryTimeRanges, meetingDurationMinutes);
+    List<TimeRange> availableOptionalTimeRanges =
+        getAvailableTimeRanges(unavailableOptionalTimeRanges, meetingDurationMinutes);
+
+    // If no mandatory attendees are passed, return the available TimeRanges for optional attendees
+    // else if no optional attendees are passed, or no TimeRanges are available for optional
+    // attendees then return available TimeRanges for mandatory attendees, else merge the available
+    // TimeRanges for mandatory attendees with available TimeRanges for optional employees.
+    if (mandatoryAttendees.isEmpty()) {
+      return availableOptionalTimeRanges;
+    } else if (optionalAttendees.isEmpty() || availableOptionalTimeRanges.isEmpty()) {
+      return availableMandatoryTimeRanges;
+    } else {
+      return mergeAvailableTimeRanges(
+          availableMandatoryTimeRanges, availableOptionalTimeRanges, meetingDurationMinutes);
+    }
+  }
+
+  /** Returns a List of TimeRanges that are unavailable in as little TimeRanges as possible. */
+  private List<TimeRange> getUnavailableTimeRanges(
+      Collection<Event> events, Collection<String> attendees) {
+    List<TimeRange> unavailableTimeRanges = new ArrayList<>();
+
+    // If an event's attendee list contains a names of an attendee that needs to attend
+    // (from Collection<String> attendees, then we add the event's TimeRange to a list.
+    events.forEach(
+        event -> {
+          Collection<String> eventAttendees = event.getAttendees();
+
+          for (String attendee : eventAttendees) {
+            if (!attendees.contains(attendee)) {
```

Comment:
Since `attendees` is a `Collection`, for each attendee in the event you're going to search through the entire `Collection` of attendees to find a match which is and O(n) operation, but if you first convert `attendees` to a `Set` (and use the same `Set` for every iteration in this loop) you reduce this check O(1).

Reply:
Done.